examples/frameworks/flaskapp_gevent_kafka.py

```diff
@@ -0,0 +1,53 @@
+# Run with:
+#
+#   $ gunicorn flaskapp:app
+#
+#   pip install kafka-python gevent
+#
+###############################
+# Author : mohammed.vaghjipurwala@gmail.com
+# Version : 1.0
+# Description : Implementation of Kafka Consumer & Producer with
+#               Flask & gunicorn with gevent
+#
+# Blog: https://medium.com/@mohammed.vaghy/implementing-kafka-with-gunicorn-gevent-workers-87e679926ddc
+#
+###############################
+
+from flask import Flask
+from gevent import monkey
+from json import loads, dumps
+from kafka import KafkaConsumer, KafkaProducer
+
+app = Flask(__name__)
+
+monkey.patch_all(socket=False, thread=False)
```

Comment:
I would be reluctant to recommend passing any non-default arguments to `monkey.patch_all()`.

You write:
>  I found that Gevent blocks consumers from reading messages continuously. So Monkey patching Gevent with disabling socket & thread.

Can you elaborate more on what you saw? That's not how gevent should behave, at all. Something like that is often a sign of monkey-patching too late, or a conflict with native code. (Here, we're definitely patching later than recommended as there are numerous imports before the patch; that's a [general problem in gunicorn too](https://github.com/benoitc/gunicorn/issues/1566), at least in some configurations.) Does gevent's [blocking monitor](http://www.gevent.org/monitoring.html#blocking) report anything when you see this behaviour?

Reply:
Hi @jamadden , 

I'm working on a product that fetches near to realtime data(statistics) from the server/3rd party tools, parse the input data, run some AI analytics on it and provide insights & solutions.

For real-time data transmission between the different services, we are using Kafka messaging broker.
When we were running in the sync mode the transmission of data was flawless. However, for performance benefits, we added Gevent as our worker on Gunicorn.
Post this, we started seeing that the consumer read only one message and then stopped and moved to the hanged state.

After some research, we found that the Gevent was blocking the consumer thread to read the messages continuously. Hence, we implemented Monkey patching disabling socket & thread.

This solution worked like a charm and resolved this issue permanently !!!

Reply:
I'm sorry, I wasn't clear. I'm the maintainer of gevent, and I wasn't asking for details about your system, I was asking for details about your research and conclusion. Exactly *how* did you determine that "that the Gevent was blocking the consumer thread to read the messages continuously" and *why* was it that "disabling socket & thread" was a solution? As I said, the behaviour you described is not something that gevent should ever do; if it does, it's a misconfiguration, an incompatibility, or a bug. So I'm trying to understand which of those it is, and exactly what's going on.

Disabling monkey-patching of threads and sockets should rarely ever be a solution, and I don't want to recommend it without knowing a lot more than we do — that's almost equivalent to simply not monkey-patching at all. Those are usually the two most important things to monkey-patch in existing code; without those, all network operations will be blocking and things usually just get worse. (Unless, of course, you're directly using gevent APIs like `gevent.socket`, but that's not in this example.) 

Besides, when your WSGI app is imported into gunicorn, gunicorn is going to run `monkey.patch_all()` *again*, and that's going to patch, you guessed it, threads and sockets. (Most likely with a warning about using `patch_all` twice with different arguments.)

As a guess, a problem like you described can happen when there is native code (a C extension) involved that directly makes blocking operating system calls; gevent cannot patch those, and they don't cooperate with the event loop, so things might appear to hang. Putting the calls to that code into a native thread can solve the problem. But the way to do that is usually not to avoid monkey-patching thread, it's to use the threadpool, and/or queues and/or async watchers. Is there any evidence that there is native code doing the blocking?

Reply:
Apologies for misunderstanding you query, The solution was implemented in the product almost 5-7 months back. Hence I donot have the exact logs at that moment.However, from what i recall we added debug loggers at kafka level first and then got to know that the after reading the first message it moved to the hang state it was able to get the heart_beat but no messages were consumed.
I forgot to add on that we were using async threads for starting multiple Consumers
After this, we tried in sync mode again and it worked without any glitches.
Then, we checked on the internet for why is it being blocked. we found few articles from which we assummed that gevent might be blocking the consumer thread.
So, we started the research to resolve the blocking issue and found some url which mentioned using `monkey.patch_all(socket=False, thread=False)`. This is how the issue got solved. Later we never tried changing this settings.

we then switched the async threads with greenlets without removing the patch_all and we are not seeing the error of consumer getting hanged.

As per your recommendation, we did not check for any blocking OS calls however, I would try again without monkey patching since we are already using the using greenlets.

Reply:
@jamadden 

Apologies for the delay. As per your recommendation, I have removed the monkey patch from the code as we already have the greenlets for async processing.
I have updated the file with the above changes.

Could you please review and merge. 

Thanks